File: SnakeDesktop/Mono/Renderer.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using SnakeModel;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SnakeDesktop.Mono
{
    class Renderer
    {
        private SnakeGame game;

        private float scale;
        private Texture2D head;
        private Texture2D body;
        private Texture2D background;
        private Texture2D food;
        private SpriteFont main;

        private int paddingX;
        private int paddingY;

        public Renderer(SnakeGame game, GraphicsDevice gd, SpriteFont main)
        {
            this.game = game;

            scale = ((Math.Min(gd.Viewport.Width, gd.Viewport.Height) / (float)SnakeGame.SIZE)) * (float)Math.Cos(Math.PI/4); //For round watch faces

            paddingX = (int)((gd.Viewport.Width - scale * SnakeGame.SIZE) / 2);
            paddingY = (int)((gd.Viewport.Height - scale * SnakeGame.SIZE) / 2);

            head = new Texture2D(gd, 1, 1);
            head.SetData<Color>(new Color[] { Color.DarkOrange });

            body = new Texture2D(gd, 1, 1);
            body.SetData<Color>(new Color[] { new Color(70,70,70) });

            background = new Texture2D(gd, 1, 1);
            background.SetData<Color>(new Color[] { Color.White });

            food = new Texture2D(gd, 1, 1);
            food.SetData<Color>(new Color[] { Color.Cyan });

            this.main = main;
        }

        public void Render(SpriteBatch sb)
        {
            sb.Begin();

            if (!game.GameOver)
            {
                DrawBackground(new Color(16, 16, 16), sb);
                sb.Draw(head, new Rectangle((int)(game.Snake.Head.X * scale) + paddingX, (int)(game.Snake.Head.Y * scale) + paddingY, (int)(scale), (int)scale), null, Color.White);
                foreach (SnekPiece piece in game.Snake.Pieces)
                {
                    DrawSnekPiece(piece, sb);
                }
                sb.Draw(food, new Rectangle((int)(game.FoodX * scale) + paddingX, (int)(game.FoodY * scale) + paddingY, (int)(scale), (int)scale), null, Color.White);
            }
            else
            {
                ShowGameOver(sb);
            }

            DrawScore(sb);

            sb.End();
        }

        public int GetGameX(int x)
        {
            return (int)((x - paddingX) / scale);
        }

        public int GetGameY(int y)
        {
            return (int)((y - paddingY) / scale);
        }

        private void DrawSnekPiece(SnekPiece piece, SpriteBatch sb)
        {
            sb.Draw(body, new Rectangle((int)(piece.X*scale) + paddingX, (int)(piece.Y*scale) + paddingY, (int)(scale), (int)scale), null, Color.White);
        }

        private void DrawScore(SpriteBatch sb)
        {
            sb.DrawString(main, "Highest: 000", new Vector2(paddingX + (SnakeGame.SIZE * scale / 2) - scale * 5, scale - (float)(Math.Cos(Math.PI/2)*scale/2)), Color.Wheat);
            sb.DrawString(main, "Current: " + GetScoreString(), new Vector2(paddingX + (SnakeGame.SIZE * scale / 2) - scale * 5, paddingY + scale*SnakeGame.SIZE + (scale - (float)(Math.Cos(Math.PI / 2) * scale / 2))), Color.Wheat);
        }

        private string GetScoreString()
        {
            return (1000 + game.Score).ToString().Substring(1);
        }

        private void DrawBackground(Color color, SpriteBatch sb)
        {
            sb.Draw(background, new Rectangle((int)(0 * scale) + paddingX, (int)(0 * scale) + paddingY, (int)(SnakeGame.SIZE * scale), (int)(SnakeGame.SIZE * scale)), null, color);
        }

        private void ShowGameOver(SpriteBatch sb)
        {
            DrawBackground(Color.DarkOrange, sb);
            sb.DrawString(main, "Game over!", new Vector2(paddingX + (SnakeGame.SIZE * scale / 2) - scale * 5, paddingY + (SnakeGame.SIZE * scale / 2) - 2*scale), Color.Black);
        }
    }
}

File: java/src/desktop/Renderer.java
package desktop;

import model.SnakeGame;
import model.SnekPiece;

import javax.swing.*;
import java.awt.*;
import java.awt.event.KeyListener;
import java.awt.event.MouseListener;

public class Renderer {

    public static final int WIDTH = 500;
    public static final int HEIGHT = 500;

    private SnakeGame game;

    private double scale;
    private Color head;
    private Color body;
    private Color background;
    private Color text;
    private Color food;
    private Font main;

    private JFrame frame;
    private JPanel surface;

    private int paddingX;
    private int paddingY;

    public Renderer(SnakeGame game) {
        this.game = game;

        scale = ((Math.min(WIDTH, HEIGHT) / (double) SnakeGame.SIZE)) * Math.cos(Math.PI / 4); //For round watch faces

        paddingX = (int) ((WIDTH - scale * SnakeGame.SIZE) / 2);
        paddingY = (int) ((HEIGHT - scale * SnakeGame.SIZE) / 2);

        head = new Color(161, 71, 0);
        body = new Color(100, 100, 100);
        background = Color.white;
        food = Color.cyan;
        text = new Color(245, 222, 179);

        main = new Font("ArialBlack", Font.BOLD, 20);
    }

    public void render(Graphics2D g) {
        g.setColor(Color.black);
        g.fillRect(0,0,WIDTH, HEIGHT);
        if (!game.isGameOver()) {
            drawBackground(new Color(16, 16, 16), g);

            drawHead(g);
            for (SnekPiece piece : game.getSnake().getBody()) {
                drawSnekPiece(piece, g);
            }

            g.setColor(food);
            g.fillRect(
                    (int) (game.getFoodX() * scale) + paddingX,
                    (int) (game.getFoodY() * scale) + paddingY,
                    (int) (scale), (int) scale
            );
        } else {
            showGameOver(g);
        }

        drawScore(g);
    }

    public int getGameX(int x) {
        return (int) ((x - paddingX) / scale);
    }

    public int getGameY(int y) {
        return (int) ((y - paddingY) / scale);
    }

    public void start(MouseListener listener) {
        frame = new JFrame("Snake for Wear (but now on desktop)");
        frame.getContentPane().setPreferredSize(new Dimension(500, 500));
        frame.pack();
        frame.setLocationRelativeTo(null);
        frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);

        surface = new JPanel() {
            @Override
            public void paintComponent(Graphics graphics) {
                render((Graphics2D) graphics);
            }
        };
        surface.setFocusable(true);
        surface.setBackground(Color.black);

        surface.addMouseListener(listener);
        frame.add(surface);

        frame.setVisible(true);
    }

    public void update() {
        surface.repaint();
    }

    private void drawHead(Graphics2D g) {
        g.setColor(head);
        g.fillRect(
                (int) Math.round(game.getSnake().getHead().getX() * scale) + paddingX,
                (int) Math.round(game.getSnake().getHead().getY() * scale) + paddingY,
                (int) Math.round(scale), (int) Math.round(scale)
        );
    }

    private void drawSnekPiece(SnekPiece piece, Graphics2D g) {
        g.setColor(body);
        g.fillRect(
                (int) Math.round(piece.getX() * scale) + paddingX,
                (int) Math.round(piece.getY() * scale) + paddingY,
                (int)Math.round(scale), (int)Math.round(scale)
        );
    }

    private void drawScore(Graphics2D g) {
        g.setFont(main);
        g.setColor(text);
        String highest = "Highest: 000";
        g.drawString(highest,
                WIDTH / 2 - g.getFontMetrics().stringWidth(highest) / 2,
                (int) paddingY - g.getFontMetrics().getHeight() // (scale - (Math.cos(Math.PI / 2) * scale / 2))
        );

        String current = "Current: " + getScoreString();
        g.drawString(current,
                WIDTH / 2 - g.getFontMetrics().stringWidth(current) / 2,
                (int) (paddingY + scale * SnakeGame.SIZE + (scale - (Math.cos(Math.PI / 2) * scale / 2)) + g.getFontMetrics().getHeight())
        );
    }

    private String getScoreString() {
        return String.valueOf(1000 + game.getScore()).substring(1);
    }

    private void drawBackground(Color color, Graphics2D g) {
        g.setColor(color);
        int x = (int) (0 * scale) + paddingX;
        int y = (int) (0 * scale) + paddingY;
        int w = (int) (SnakeGame.SIZE * scale);
        int h = (int) (SnakeGame.SIZE * scale);
        g.fillRect(x,y,w,h);
        g.setColor(head);
        g.drawRect(x,y,w,h);
    }

    private void showGameOver(Graphics2D g) {
        drawBackground(head, g);
        g.setFont(main);
        g.drawString(
                "Game over!",
                (int) (paddingX + (SnakeGame.SIZE * scale / 2) - scale * 5),
                (int) (paddingY + (SnakeGame.SIZE * scale / 2) - 2 * scale)
        );
    }
}